core/src/main/java/io/questdb/griffin/engine/QueryProgress.java

```diff
@@ -136,10 +141,15 @@ public static void logEnd(
         // as well as already converted to an immutable String, as needed to queue it up for handling
         // at a later time. For this reason, do not assign queryTrace.queryText = sqlText here.
         if (queryTrace != null && engine.getConfiguration().isQueryTracingEnabled()) {
+            if (queryTrace.shouldSubstituteBindVariables(engine.getConfiguration())) {
+                // unless we need to change it!
+                queryTrace.queryText = substituteBindVariablesIntoQueryText(sqlText, executionContext);
```

Comment:
could we store the `sqlText`  in a sink in `QueryTrace` until we get here?
now we create a string in the constructor of `QueryProgress` to set `queryTrace.queryText`, then possibly another one here to replace it, and that string just gets thrown away.
if QueryTrace had a sink instead, we could generate only one string per query.
or maybe even better, we could try not to set queryText in the constructor of `QueryProgress`.
it would be set in `logEnd()`/`logError()`.